File: python/demos/demo_overview.py
"""demo_overview.py

Quick tour of the Music Perception Toolbox (mpt) for Python.
Sections follow the same order as the User Guide.

Uses the son clave rhythm [0, 3, 6, 10, 12] in a 16-step cycle and
the diatonic scale [0, 2, 4, 5, 7, 9, 11] in 12-EDO as running
examples.
"""

import numpy as np
import mpt

np.set_printoptions(precision=3, suppress=True)

# ===================================================================
#  1. Pitch / frequency conversion  (User Guide §6.6)
# ===================================================================

print("=== Pitch conversion ===")
print(f"  MIDI 60 = {mpt.convert_pitch(60, 'midi', 'hz'):.2f} Hz")
print(f"  440 Hz  = {mpt.convert_pitch(440, 'hz', 'cents'):.0f} cents")
print(f"  440 Hz  = {mpt.convert_pitch(440, 'hz', 'erb'):.2f} ERB-rate")

# ===================================================================
#  2. Spectral enrichment  (User Guide §6.2)
# ===================================================================

print("\n=== Spectral enrichment ===")
chord = np.array([0.0, 400.0, 700.0])
p, w = mpt.add_spectra(chord, None, "harmonic", 8, "powerlaw", 1)
print(f"  3 pitches × 8 harmonics = {len(p)} partials")
print(f"  First 8 offsets: {np.round(p[:8], 1)}")

# ===================================================================
#  3. Expectation tensors and SPCS  (User Guide §6.1, §3.3)
# ===================================================================

print("\n=== Spectral pitch class similarity ===")
scale_cents = [0, 200, 400, 500, 700, 900, 1100]
major = [0, 400, 700]
minor = [0, 300, 700]
dim_t = [0, 300, 600]

# Add harmonic spectra (24 partials, 1/n rolloff)
scale_p, scale_w = mpt.add_spectra(scale_cents, None, 'harmonic', 24, 'powerlaw', 1)

for name, chord_cents in [("Major", major), ("Minor", minor), ("Dim", dim_t)]:
    chord_p, chord_w = mpt.add_spectra(chord_cents, None, 'harmonic', 24, 'powerlaw', 1)
    s = mpt.cos_sim_exp_tens_raw(
        scale_p, scale_w, chord_p, chord_w,
        10, 1, False, True, 1200, verbose=False
    )
    print(f"  Diatonic vs {name:5s} triad: {s:.3f}")

# ===================================================================
#  4. Consonance and harmonicity  (User Guide §6.3)
# ===================================================================

print("\n=== Harmonicity and entropy (JI major triad) ===")
ji_triad = [0, 386.31, 701.96]
spec = ["harmonic", 24, "powerlaw", 1]

hMax, hEnt = mpt.template_harmonicity(ji_triad, None, 12, chord_spectrum=spec)
print(f"  Template harmonicity (hMax):     {hMax:.4f}")
print(f"  Template harmonicity (hEntropy): {hEnt:.4f}")

h = mpt.tensor_harmonicity(ji_triad, None, 12, spectrum=spec)
print(f"  Tensor harmonicity:              {h:.4f}")

H = mpt.spectral_entropy(ji_triad, None, 12, spectrum=spec)
print(f"  Spectral entropy:                {H:.4f}")

print("\n=== JI vs 12-EDO comparison ===")
edo_triad = [0, 400, 700]
for name, triad in [("JI", ji_triad), ("12-EDO", edo_triad)]:
    hMax, _ = mpt.template_harmonicity(triad, None, 12, chord_spectrum=spec)
    H = mpt.spectral_entropy(triad, None, 12, spectrum=spec)
    print(f"  {name:6s}  hMax={hMax:.4f}  specEntropy={H:.4f}")

print("\n=== Roughness ===")
p_cents = mpt.convert_pitch([60, 64, 67], "midi", "cents")
p_r, w_r = mpt.add_spectra(p_cents, None, "harmonic", 8, "powerlaw", 1)
f_hz = mpt.convert_pitch(p_r, "cents", "hz")
r = mpt.roughness(f_hz, w_r)
print(f"  C major triad (8 harmonics): roughness = {r:.4f}")

# ===================================================================
#  5. Balance and evenness  (User Guide §6.4)
# ===================================================================

print("\n=== Balance and evenness ===")

print("  Diatonic scale [0,2,4,5,7,9,11] in 12-EDO:")
diat = [0, 2, 4, 5, 7, 9, 11]
b = mpt.balance(diat, None, 12)
e = mpt.evenness(diat, 12)
print(f"    Balance:  {b:.3f}")
print(f"    Evenness: {e:.3f}")

print("  Son clave [0,3,6,10,12] in 16:")
clave = [0, 3, 6, 10, 12]
b = mpt.balance(clave, None, 16)
e = mpt.evenness(clave, 16)
print(f"    Balance:  {b:.3f}")
print(f"    Evenness: {e:.3f}")

# ===================================================================
#  6. Scale and rhythm structure  (User Guide §6.5)
# ===================================================================

print("\n=== Scale structure (diatonic) ===")
c, nc = mpt.coherence(diat, 12)
sq, nd = mpt.sameness(diat, 12)
print(f"  Coherence: {c:.3f} ({nc} failure)")
print(f"  Sameness:  {sq:.3f} ({nd} ambiguity)")

H1, _ = mpt.n_tuple_entropy(diat, 12, 1)
H2, _ = mpt.n_tuple_entropy(diat, 12, 2)
print(f"  1-tuple entropy: {H1:.3f}")
print(f"  2-tuple entropy: {H2:.3f}")

print("\n=== Rhythm structure (son clave) ===")
c, nc = mpt.coherence(clave, 16)
sq, nd = mpt.sameness(clave, 16)
print(f"  Coherence: {c:.3f} ({nc} failures)")
print(f"  Sameness:  {sq:.3f} ({nd} ambiguities)")

H1, _ = mpt.n_tuple_entropy(clave, 16, 1)
H2, _ = mpt.n_tuple_entropy(clave, 16, 2)
print(f"  1-tuple entropy: {H1:.3f}")
print(f"  2-tuple entropy: {H2:.3f}")

h = mpt.mean_offset(clave, None, 16)
print(f"  Mean offset: {np.round(h, 3)}")

edg, _ = mpt.edges(clave, None, 16)
print(f"  Edges: {np.round(edg, 3)}")

y = mpt.markov_s(clave, None, 16)
print(f"  Markov(3): {np.round(y, 3)}")

print("\nDone.")

File: matlab/demos/demo_overview.m
%% demo_overview.m
%  Quick tour of the Music Perception Toolbox.
%  Sections follow the same order as the User Guide.
%
%  Uses the son clave rhythm [0, 3, 6, 10, 12] in a 16-step cycle and
%  the diatonic scale [0, 2, 4, 5, 7, 9, 11] in 12-EDO as running
%  examples.
%
%  Uses: convertPitch, addSpectra, cosSimExpTens, templateHarmonicity,
%        tensorHarmonicity, spectralEntropy, roughness, balanceCircular,
%        evennessCircular, coherence, sameness, nTupleEntropy, meanOffset,
%        edges, markovS
%  (from the Music Perception Toolbox).

%% === 1. Pitch / frequency conversion (User Guide §6.6) ===

fprintf('=== Pitch conversion ===\n');
fprintf('  MIDI 60 = %.2f Hz\n', convertPitch(60, 'midi', 'hz'));
fprintf('  440 Hz  = %.0f cents\n', convertPitch(440, 'hz', 'cents'));
fprintf('  440 Hz  = %.2f ERB-rate\n', convertPitch(440, 'hz', 'erb'));

%% === 2. Spectral enrichment (User Guide §6.2) ===

fprintf('\n=== Spectral enrichment ===\n');
chord = [0; 400; 700];
[p, w] = addSpectra(chord, [], 'harmonic', 8, 'powerlaw', 1);
fprintf('  3 pitches × 8 harmonics = %d partials\n', numel(p));
fprintf('  First 8 offsets: %s\n', mat2str(round(p(1:8)', 1)));

%% === 3. Expectation tensors and SPCS (User Guide §6.1, §3.3) ===

fprintf('\n=== Spectral pitch class similarity ===\n');
scale_cents = [0, 200, 400, 500, 700, 900, 1100];
chords = {[0, 400, 700], [0, 300, 700], [0, 300, 600]};
chordNames = {'Major', 'Minor', 'Dim'};

% Add harmonic spectra (24 partials, 1/n rolloff)
[scale_p, scale_w] = addSpectra(scale_cents, [], 'harmonic', 24, 'powerlaw', 1);

for i = 1:numel(chords)
    [chord_p, chord_w] = addSpectra(chords{i}, [], 'harmonic', 24, 'powerlaw', 1);
    s = cosSimExpTens(scale_p, scale_w, chord_p, chord_w, ...
                      10, 1, false, true, 1200, 'verbose', false);
    fprintf('  Diatonic vs %-5s triad: %.3f\n', chordNames{i}, s);
end

%% === 4. Consonance and harmonicity (User Guide §6.3) ===

fprintf('\n=== Harmonicity and entropy (JI major triad) ===\n');
ji_triad = [0, 386.31, 701.96];
spec = {'harmonic', 24, 'powerlaw', 1};

[hMax, hEnt] = templateHarmonicity(ji_triad, [], 12, ...
    'chordSpectrum', spec);
fprintf('  Template harmonicity (hMax):     %.4f\n', hMax);
fprintf('  Template harmonicity (hEntropy): %.4f\n', hEnt);

h = tensorHarmonicity(ji_triad, [], 12, 'spectrum', spec);
fprintf('  Tensor harmonicity:              %.4f\n', h);

H = spectralEntropy(ji_triad, [], 12, 'spectrum', spec);
fprintf('  Spectral entropy:                %.4f\n', H);

fprintf('\n=== JI vs 12-EDO comparison ===\n');
edo_triad = [0, 400, 700];
for name = {"JI", "12-EDO"}
    if strcmp(name{1}, 'JI')
        triad = ji_triad;
    else
        triad = edo_triad;
    end
    [hMax, ~] = templateHarmonicity(triad, [], 12, 'chordSpectrum', spec);
    H = spectralEntropy(triad, [], 12, 'spectrum', spec);
    fprintf('  %-6s  hMax=%.4f  specEntropy=%.4f\n', name{1}, hMax, H);
end

fprintf('\n=== Roughness ===\n');
p_cents = convertPitch([60, 64, 67], 'midi', 'cents');
[p_r, w_r] = addSpectra(p_cents, [], 'harmonic', 8, 'powerlaw', 1);
f_hz = convertPitch(p_r, 'cents', 'hz');
r = roughness(f_hz, w_r);
fprintf('  C major triad (8 harmonics): roughness = %.4f\n', r);

%% === 5. Balance and evenness (User Guide §6.4) ===

fprintf('\n=== Balance and evenness ===\n');

diat = [0, 2, 4, 5, 7, 9, 11];
fprintf('  Diatonic scale [0,2,4,5,7,9,11] in 12-EDO:\n');
fprintf('    Balance:  %.3f\n', balanceCircular(diat, [], 12));
fprintf('    Evenness: %.3f\n', evennessCircular(diat, 12));

clave = [0, 3, 6, 10, 12];
fprintf('  Son clave [0,3,6,10,12] in 16:\n');
fprintf('    Balance:  %.3f\n', balanceCircular(clave, [], 16));
fprintf('    Evenness: %.3f\n', evennessCircular(clave, 16));

%% === 6. Scale and rhythm structure (User Guide §6.5) ===

fprintf('\n=== Scale structure (diatonic) ===\n');
[c, nc] = coherence(diat, 12);
[sq, nd] = sameness(diat, 12);
fprintf('  Coherence: %.3f (%d failure)\n', c, nc);
fprintf('  Sameness:  %.3f (%d ambiguity)\n', sq, nd);

H1 = nTupleEntropy(diat, 12, 1);
H2 = nTupleEntropy(diat, 12, 2);
fprintf('  1-tuple entropy: %.3f\n', H1);
fprintf('  2-tuple entropy: %.3f\n', H2);

fprintf('\n=== Rhythm structure (son clave) ===\n');
[c, nc] = coherence(clave, 16);
[sq, nd] = sameness(clave, 16);
fprintf('  Coherence: %.3f (%d failures)\n', c, nc);
fprintf('  Sameness:  %.3f (%d ambiguities)\n', sq, nd);

H1 = nTupleEntropy(clave, 16, 1);
H2 = nTupleEntropy(clave, 16, 2);
fprintf('  1-tuple entropy: %.3f\n', H1);
fprintf('  2-tuple entropy: %.3f\n', H2);

h = meanOffset(clave, [], 16);
fprintf('  Mean offset: %s\n', mat2str(round(h, 3)));

edg = edges(clave, [], 16);
fprintf('  Edges: %s\n', mat2str(round(edg, 3)));

y = markovS(clave, [], 16);
fprintf('  Markov(3): %s\n', mat2str(round(y, 3)));

fprintf('\nDone.\n');
